profiler/include/profiler/profile_gemm_universal_impl.hpp

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+#include <iomanip>
+#include <iostream>
+#include <typeinfo>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
+#include "ck/tensor_operation/gpu/device/impl/device_gemm_xdl_cshuffle_v3.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+
+#include "ck/library/tensor_operation_instance/gpu/gemm_universal.hpp"
+
+#include "ck/library/utility/check_err.hpp"
+#include "ck/library/utility/device_memory.hpp"
+#include "ck/library/utility/host_tensor.hpp"
+#include "ck/library/utility/host_tensor_generator.hpp"
+#include "ck/library/utility/literals.hpp"
+#include "ck/library/reference_tensor_operation/cpu/reference_gemm.hpp"
+
+namespace ck {
+namespace profiler {
+
+template <typename ADataType,
+          typename BDataType,
+          typename AccDataType,
+          typename CDataType,
+          typename ALayout,
+          typename BLayout,
+          typename CLayout>
+bool profile_gemm_universal_impl(int do_verification,
+                                 int init_method,
+                                 bool do_log,
+                                 bool time_kernel,
+                                 int M,
+                                 int N,
+                                 int K,
+                                 int StrideA,
+                                 int StrideB,
+                                 int StrideC,
+                                 int KBatch,
+                                 int n_warmup,
+                                 int n_iter)
+{
+    bool pass = true;
+
+    auto f_host_tensor_descriptor =
+        [](std::size_t row, std::size_t col, std::size_t stride, auto layout) {
+            using namespace ck::literals;
+
+            if(is_same<decltype(layout), tensor_layout::gemm::RowMajor>::value)
+            {
+                return HostTensorDescriptor({row, col}, {stride, 1_uz});
+            }
+            else
+            {
+                return HostTensorDescriptor({row, col}, {1_uz, stride});
+            }
+        };
+
+    Tensor<ADataType> a_m_k(f_host_tensor_descriptor(M, K, StrideA, ALayout{}));
+    Tensor<BDataType> b_k_n(f_host_tensor_descriptor(K, N, StrideB, BLayout{}));
+    Tensor<CDataType> c_m_n_host_result(f_host_tensor_descriptor(M, N, StrideC, CLayout{}));
+    Tensor<CDataType> c_m_n_device_result(f_host_tensor_descriptor(M, N, StrideC, CLayout{}));
+
+    std::cout << "a_m_k: " << a_m_k.mDesc << std::endl;
+    std::cout << "b_k_n: " << b_k_n.mDesc << std::endl;
+    std::cout << "c_m_n: " << c_m_n_device_result.mDesc << std::endl;
+
+    switch(init_method)
+    {
+    case 0: break;
+    case 1:
+        a_m_k.GenerateTensorValue(GeneratorTensor_2<ADataType>{-1, 2});
```

Comment:
IIUC, this could generate `nan` for `f8_t`. Since `f8_t` is not a valid floating point data type, the `static_cast` in `GeneratorTensor_2` only performs integer trunc. If `-1` is generated, it will be `0b11111111`, which is `nan`.

Similarly, if it generates `1` and `2`, it's gonna be subnormal values.

Reply:
Interesting finding👍 , let me check it.

Reply:
This could potentially cause test failures because the HW might handle `nan` and subnormal values in different ways from host.

Reply:
Hi @shiltian ,
f8_t random integer generation type convert not through static_cast but a ck-defined type_convert function.
do you observe Nan generated in your experiment?
https://github.com/ROCm/composable_kernel/blob/4ec5c52a0c01e9b34ae9c3918a0c9372075e5852/library/include/ck/library/utility/host_tensor_generator.hpp#L124-L138

Reply:
Oh okay, I didn't notice this. Thanks for pointing it out. Well, I did notice nan but it might be irrelevant.

Reply:
Hi there, I indeed notice `nan` generated in my test.

```
Running main() from /root/build/_deps/gtest-src/googletest/src/gtest_main.cc
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from TestGemmUniversal_MK_KN/0, where TypeParam = St5tupleIJDB8_DF16_DF16_EE
[ RUN      ] TestGemmUniversal_MK_KN/0.Regular
a_m_k: dim 2, lengths {512, 512}, strides {512, 1}
b_k_n: dim 2, lengths {512, 512}, strides {512, 1}
c_m_n: dim 2, lengths {512, 512}, strides {512, 1}
rotating count: 1
found 88 instances
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 128x128x64, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v4, BlkGemmPipelinePrefetchStages: 4, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 256x128x128, WaveTile: 32x32, WaveMap: 4x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v3, BlkGemmPipelinePrefetchStages: 2, KBatch 1
DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 192x256x64, WaveTile: 32x32, WaveMap: 3x4, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v3, BlkGemmPipelinePrefetchStages: 2 does not support this problem
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 128x128x128, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v3, BlkGemmPipelinePrefetchStages: 2, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 128x128x64, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v3, BlkGemmPipelinePrefetchStages: 2, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 128x128x64, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x4, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v5, BlkGemmPipelinePrefetchStages: 3, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<Default, RRR> BlkSize: 256, BlkTile: 128x128x64, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Interwave, BlkGemmPipelineVersion: v1, BlkGemmPipelinePrefetchStages: 1, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
Perf:          0 ms, inf TFlops, inf GB/s, DeviceGemmXdlUniversal<KPadding, RRR> BlkSize: 256, BlkTile: 128x128x64, WaveTile: 32x32, WaveMap: 2x2, VmemReadVec: 16x8, BlkGemmPipelineScheduler: Intrawave, BlkGemmPipelineVersion: v4, BlkGemmPipelinePrefetchStages: 4, KBatch 1
Error: Incorrect results!        out[0] != ref[0]: nan != nan
Error: Incorrect results!        out[1] != ref[1]: nan != nan
Error: Incorrect results!        out[2] != ref[2]: nan != nan
Error: Incorrect results!        out[3] != ref[3]: nan != nan
max err: -9.632111e-05, number of errors: 262144, 100% wrong values
...
```

Reply:
Hi @shiltian ,
Thanks for the report. Could you provide steps to reproduce?

Reply:
We can continue discussion in SWDEV-476494.